[sgen] Optimize LOS for better locality and parallelization.

lateralusX opened the pull request:
Running parallel scan in minor GC doesn't scale out well over multiple cores when the LOS list is getting big. One of the reasons why this happens is the way each job travers the LOS list. Currently, all jobs scan the complete LOS list, skipping jobs not matching their individual job index.

Problem with this is that each thread needs to touch every element in the list, and by default, we  create number of cores * 4 jobs scanning LOS objects. That means we will touch every element 32 times on an 8 core machine, moving a lot of memory between cache levels, increasing the total scan time for the complete LOS scan job.

This is especially problematic on hardware with slower memory bus and smaller caches, but with many cores. On such hardware the accumulated scan time can be bigger when parallelize compared to just having one thread doing the scan, but all depends on how many objects each job needs to scan. In one scenario we had a major heap of 3 GB and ~600 MB of LOS , split on ~ 12500 LOS objects. Doing minor GC on that heap took ~20 ms running without parallelization. With default parallelization we were down to ~19 ms and the time scanning the LOS objects didn't improve. NOTE, the more work each job gets (more memory to scan) we will get a better default parallelization, but still a lot of room for improvement.

Changing from a normal linked list to a SgenArrayList makes it more effective to parallelize the scan jobs and since the list is using sequential memory, this will also reduce cache misses for each thread iterating LOS objects. Tagging objects in SgenArrayList with reference information reduce cache misses dramatically in cases where objects not including references exists on the list, since scanning jobs can skip these items without any need to touch object memory causing cache misses.

Same scenario running with this fix gives us minor GC pause times, without parallelization in ~ 7ms (down from ~ 20ms) and with parallelization, < 5ms (down from 19ms). NOTE, the more work each job gets (more memory to scan) the better result using parallelization, so the delta will increase between using and not using parallelization, this scenario primarily stress the extreme case where most LOS objects doesn't have any references.

In total, this optimization reduce minor GC pause time from ~ 19ms when using parallelization down to < 5ms, that's close to 4x improvement in minor GC pause times.

This is the first PR in a series of optimizations around minor GC parallelization that have made a significant performance increase, both on local tests as well as in real world large applications very sensitive to minor GC pause times, like games. When applying all upcoming PR's on top of the same scenario, we are able to complete minor GC in ~ 2.5 ms using parallel minor GC, that initially took ~ 20 ms, that's 8x improvements in minor GC pause times.

lateralusX commented:
@monojenkins rebuild failed

BrzVlad commented:
Overall I have mixed feelings about this. I don't like that much that it complicates a little bit the large object list layout. It is also hard to tell how this will behave for long running programs, in theory this might need some load balancing sometimes. Also the table size is a little bit arbitrary and it would be nice if we didn't have to deal with that as a compile time constant.

I'm wondering if it is worth trying out a simpler approach. We could keep track of the number of los objects at all time (alloc++/free--). When submitting the LOS scan jobs we could instead submit a job whose job is to produce all the smaller jobs. It could pass to each job the LOSObject start and the number of objects to scan. As it traverses a section of the list, it would already enqueue a new job, so once, a small section of the list has been traversed, parallelization already starts. With this approach we keep the simple los layout, we always do balancing of the jobs and the overhead of 2 list traversals instead of one (with one also being parallelized) should be negligible (way better than 32 anyway).

The parallelization model for the los scanning is already not great, because we don't take into account the size of each object. For example scanning a 8KB object is very different than scanning an 8MB object, but we consider them all the same. Going with the approach in this PR might make it also awkward to implement that in the future. While if we have the simple job enqueuing approach, we could easily expand it to divide jobs based on total size to be scanned, instead of object count.

@lateralusX I think it is worth trying out this approach and see if it has the same benefits as the one in this PR.

jaykrell commented on mono/sgen/sgen-gc.h:
```diff
@@ -909,7 +909,39 @@ struct _LOSObject {
+#define SGEN_LOS_OBJECT_LIST_BUCKETS 31
+extern LOSObject *sgen_los_object_lists [];
+
+static inline guint
```
uint64 is a better hash type than uint32, because, in the event of hash % bucket_count colliding, if you store the full hash (not-modded) in the table also, you can compare that for equality before doing a full compare.

lateralusX replied:
This hash function is just used to pick a bucket, no need to compare against this hash. It is also using the hash function we use elsewhere, sgen_aligned_addr_hash, but could of course use a uin64 or even do the hash % bucket_count as part of this function.

jaykrell commented:
Fair enough. Comment is more for anyone implementing a hash table: 
 - hash value should be stored with full value in the table
 - hash functions should return 64 bits

So that values in the same bucket ("collision"), can be compared for equality more efficiently, by comparing their full hash value before resorting to comparing their full value.

lateralusX commented:
@BrzVlad Couple of thoughts:

Size of list could be dynamically allocated based on the job split (or configuration), if we think that the default bucket size is arbitrary (use simliar default size for other implementations, SGEN_ARRAY_LIST_BUCKETS).

I ran with this for sometime with random allocations of LOS objects and the objects seems to distribute quite well over the buckets, but of course, in theory some of the list might get more objects than other, but even if we could get bad distribution over time, it will still be split out over more list compared to what we have today. Having a larger list count compared to scan jobs would also assist that scenario a little, reducing the risk of unbalanced work loads/thread. Do we have any tests that we could use to investigate if this is something more than a potential theoretical issue?

I also thought around better split of LOS scan jobs taken into account size of objects in the distribution and if we are going to solve that, then we will need a different split, but in order to get that working we will need to know total size of LOS objects (something we probably already know), and we will need to look through the card table for each object to find out how much of that object that really needs to be scanned. Is that something we do when parallelize the major heap scan today, or could individual threads take a higher hit if they ends up with blocks that have more memory that needs to be scanned (based on marked cards)?

Regarding parallelization starting early, one thing I notice when working on this, is that none of the jobs queued in enqueue_scan_remembered_set_jobs will be picked up by thread pool until sgen_workers_start_all_workers, so currently none of the jobs can be started until we call that method since it sets the context->started to true and none of the workers will pickup jobs until that gets set.

I could benchmark alternative ways to parallelize this (like the suggestion above), but I suspect that the first traversal of the whole list (based on below observations), will take more time than expected. Did a test on targeted hardware around this. Just iterating the LOS list and calculate number of items + total size using one thread takes ~2.5 - 2.8 ms:

`Scanning LOS list calculating size: items 11102, 604015144 bytes, 2540 usecs`

The complete minor GC completes in ~6 ms with all optimizations in place. Paying 2.5 ms to use a different algorithm is a rather high price to pay. Also, that price will probably be payed more than once when we try to parallelize the LOS list in more places.

I have another PR on the way that builds on this way of splitting the LOS list into buckets, that parallelize the copy/clear of card table, that PR reduce that part from 8 ms down to 3 ms in a benchmark with 3GB major and 600 MB LOS and if we do a different algorithm, then that parallelization will be harder to accomplish since its mainly about iterating the list, comparing all found objects against card table to copy/clear marked cards, nothing will be scanned, so first iterating through the list will cost 2.5 ms in copy step and (unless we preserve the split) an additional 2.5 ms in clear, defeating pretty much what we gain in parallelize this phase in the first place.

So to summarize my thoughts:

Yes this complicates the layout of LOS list a little, but most of the code is still the same, we could even restore the old behavior by setting a size to 1 and the compiler will most likely reduce the outer loop over buckets, just leaving the inner loop, giving the same behavior as we had before. I even preserved the old way of splitting up jobs if we only have a list of size 1. If the default hard list size is an issue, we could switch into a configurable size that more easily could be tweaked.

This way of splitting the LOS list also enables for further parallelization of other tasks in minor GC, like copy/clear (mentioned above). If we change this and do the split in a different way, like running through the list once to split up the jobs dynamically as part of minor GC, then we might need to implement logic to preserve the split or we will take the hit multiple times during minor GC. If we weight in scan size as well, then the split might be different depending on what part of minor GC that we parallelize, so we might end up paying the prices more than once, and that will be a significant cost to pay and it will complicate the implementation.

The price of iterating the LOS list could be quite high, depending on hardware and number of items in the list. In the scenario I have been looking at, it's 2.5 - 2.8 ms and we are currently (with all optimizations) at minor GC pause times around ~6 ms (down from 30 ms in my local benchmark). If I stressed this a little, just allocating 8001 bytes up to a total of 600 MB LOS, then iterating through the list of 75000 LOS objects will take 14 ms, but the distribution looked OK despite the same allocation size, all lists had at least 2000 objects, most list in the range 2200 - 2500. Increasing the 

Could an alternative be to switch to configurable list size, and default to 1 (to preserve current behavior) and then have an option to increase the size meaning that we can experiment more easily with this optimization + opening up for the other PR's building on this? I also have some additional logging in later PR's that can output the split over the list on higher log levels making it possible to analyze the split over time.

BrzVlad commented:
> Is that something we do when parallelize the major heap scan today, or could individual threads take a higher hit if they ends up with blocks that have more memory that needs to be scanned (based on marked cards)?

We don't count cards to be scanned

> Regarding parallelization starting early, one thing I notice when working on this, is that none of the jobs queued in enqueue_scan_remembered_set_jobs will be picked up by thread pool until sgen_workers_start_all_workers, so currently none of the jobs can be started until we call that method since it sets the context->started to true and none of the workers will pickup jobs until that gets set.

What I was suggesting to do is to enqueue the job for computing all the smaller jobs, then start all workers. Once a fraction of the computing job is executed, new jobs that actually do the scanning will already be enqueued and executed. In this case, we wouldn't need to wait for 2-3ms.

> I have another PR on the way that builds on this way of splitting the LOS list into buckets, that parallelize the copy/clear of card table

In that case, having the additional job splitting job would be too annoying.

At this moment, I think the sanest approach would be to do exactly what we do with major blocks. We would have the same parallelization logic for both los and major blocks and might make stuff easier in the future. This would address all my concerns, while making it also easy to do the card clearing. This would mean switching from linked list for large objects to an SgenArrayList. We can then easily split any job based on indexes and LOSObject pointers will be one after the other in consecutive memory. In the future, with this approach, we might be able to completely avoid dereferencing a LOSObject if it has no cards marked, since we don't need to get the next element. Maybe this would get us even faster

lateralusX commented:
That feels quite close to what's been done in this PR and should be easy to base the other PR's on top (parallell copy/clear etc). SgenArrayList is also splitting up things into buckets and using arrays to store the actual items, so that will be something we can parallelize efficiently ontop, and it will drastically reduce the cache misses current LOS linked lists hits (when iterating the items).

I considered this approach when splitting existing linked list into buckets, but thought it might be to invasive change and just doing the split of existing list seemed like a simpler, less intrusive fix. If you are fine doing that more in-depth change around LOS objects, I'm happy to look into that option to get a fix that we all are happy with and that will reach the performance goals.

Just validating that we are on the same track, so we are thinking in splitting the LOS object header and data into two allocations, header allocated and added to SgenArrayList and data using existing allocation pattern. I quess we continue to use the LOS free list as is for the data, so pretty much we change allocation to:

BrzVlad commented:
This is not what I meant. We keep the LOSObject structure, except we will no longer need the next field, since the LOSObject will no longer be linked. Instead we would store them in a SgenArrayList (you can think of a SgenArrayList to just be an array of references, it will just be an expanding array of LOSObject*. The reason it internally contains exponentially expanding buckets is to implement lock free concurrency with adding elements to the list).

So nothing really changes from the existing implementation. Alloc = alloc_memory_for_los_object, los_objects_list.add (lso_obj), and during sweep we will *slot = NULL if the los object is dead, instead of unlinking. And we will have similar FOREACH_LOS_OBJECT defines like FOREACH_BLOCK_NO_LOCK when traversing each los object

lateralusX commented:
I updated my initial comment, so I believe we are in sync.

BrzVlad commented:
Yeah, I didn't notice the OR part. I don't think we will need an index, since we normally don't need to go from LOSObject to the location in the list, except in sweep, but in sweep we will have the slot address/index at hand since we are iterating the list.

lateralusX commented:
Yeah, just looked at sgen_los_free_object and it only seems to be called when we iterating the LOS list at the moment, so we never call this function without going from the "slot", so backpointer should not be needed.

lateralusX commented:
I will try to set a side some time to adjust this PR over to using SgenArrayList, rebase my other PR's on top of that and run all on targeted hardware to validate that we are on pair perf wise. I'm optimiztic that this could given us even more gain, since we move away from linked list altoghether, reducing cache misses. Stay tuned!

BrzVlad commented:
Consider trying out PREFETCH_CARDS for more gains.

lateralusX commented:
I experimented a little with that in the past for major scan, but it actually reduced performance at that point, but could be worth trying out for the LOS scan as well.

lateralusX commented:
Started to work on change over to SgenArrayList. I have done some initial runs on targeted hardware with all changes and it looks promising, still a little slower compared to the split list, but there is still room for additional optimization and improvements.

BrzVlad commented:
@lateralusX I forgot if you were using a real scenario or just a dummy benchmark with a lot of los objects. I'm wondering if on games they have a lot of objects with no references (maybe most of their LOS objects are just byte arrays). If it is true, we could demolish the los overhead during minors by having a separate LOSObject list for objects without references. We wouldn't need to do any traversal, card clearing or card scanning for these objects.

lateralusX commented:
What I did so far was to tag the objects in the list with an "has ref" bit (similar to what we do in major heap blocks) and will skip directly when checking pointer during iteration when finding an object without reference. That gives a rather good performance boost, but splitting in 2 different SgenArrayLists could of course reduce this even further.

BrzVlad commented:
Tagging should be fine. We can consider splitting if in a real game we have a high majority of non reference los objects and a non negligible los traversal time.

lateralusX commented:
My dummy benchmark has a lot of none reference LOS objects, but with all changes and parallelization done, the traversal of LOS objects is now starting to move of radar.

lateralusX commented:
Fixes are also benchmarked on a real large game with similar heap characteristics.

lateralusX commented:
With reference pointer tagging and some additional optimization, new solution is now (as expected) even faster than split list due to better locality. Still looking into a set of cache misses around SgenArrayList iteration to see if that could be eliminated as well.

lateralusX commented:
Updated to use SgenArrayList as LOS data structure. This improved locality, decreasing cache misses when iterating through large objects. Together with "has reference" pointer tagging the iteration times improved even more in cases where objects not including any references was included on the list, since those objects can be skipped by just looking at the pointer in the list and no need to dereference the allocated object memory, causing additional cache misses.

I also discovered an inefficient implementation under MSVC related to software emulation of __builtin_clz. Implemented corresponding intrinsic using different intrinsics under MSVC, but the most optimal one, __lzcnt, is only available on Hashwell and later for Intel CPU's, and didn't want to add dynamic cpu capability check since that would add a branch into this highly optimized inlined method, so made that into a compile time condition. By default upstream Mono will fallback to BitScanReverse, but that is still way better then the current software emulation in use. This change will also improve the iteration times of major blocks in major GC.

BrzVlad commented on mono/sgen/sgen-array-list.h:
```diff
@@ -49,14 +49,31 @@ typedef struct {
+{
+	return __lzcnt (x);
+}
+#define __builtin_clz sgen_msvc_lzcnt
```
Doesn't feel like a good idea to define __builtin_clz. We should use sgen_msvc_lzcnt in sgen, and have it defined to __builtin_clz or whatever intrinsics are present on other platforms. Maybe also have naive implementation for it on unknown platform and avoid define in sgen_array_list_index_bucket.

lateralusX replied:
Sure, I can give it a different name and move it so it could be shared within sgen, put it in sgen-gc.h?

BrzVlad replied:
It can also stay in sgen-array-list.h if not used anywhere else. I just don't want to redefine __builtin_clz which is meant to be a compiler intrinsic.

lateralusX replied:
OK, then I will keep it in sgen-array-list.h and just it a different name.

lateralusX commented:
Failures on Linux ARMv7 appears on at least two unrelated PR's.
Failures on Linux WebAssembly appears on many PR's.

lambdageek commented on mono/sgen/sgen-los.c:
```diff
@@ -432,39 +487,35 @@ static void sgen_los_unpin_object (GCObject *data);
-				mword num_cards = sgen_card_table_number_of_cards_in_range ((mword) bigobj->data, obj_size);
-				memset (bigobj->cardtable_mod_union, 0, num_cards);
+	FOREACH_LOS_OBJECT_NO_LOCK (obj) {
+		SGEN_ASSERT (0, !SGEN_OBJECT_IS_PINNED (obj->data), "Who pinned a LOS object?");
```
Can a user do this? (Users can't tell a large object apart from a small object, right?)  Or does Sgen just ignore requests to pin/unpin large objects?

BrzVlad replied:
This pinning is internal to the GC and it is only set during a collection. When the mutator wants to pin an object it allocates a pinned gchandle, which means the reference to that object is stored in an sgen structure that is internally registered as a conservative root (see ROOT_TYPE_PINNED coupled with MONO_ROOT_SOURCE_GC_HANDLE). During the collection the GC will pin all objects encountered there and make sure they are unpinned when the collection is done. The objects will remain in the other structure until the user deregisters the gchandle.

BrzVlad replied:
I don't think we ignore request to pin LOS objects, even though we technically could.

lateralusX replied:
Found this in sgen_alloc_obj_pinned:

BrzVlad commented on mono/sgen/sgen-los.c:
```diff
@@ -726,6 +799,32 @@ sgen_los_pin_object (GCObject *data)
 }
 
+void
+sgen_los_iterate_pin_objects (SgenGrayQueue *gray_queue, gboolean finish_concurrent_mode)
```
This name is confusing, I thought we iterated pinned objects. It should be `sgen_los_pin_objects`

lateralusX replied:
Agree, intention wit the naming was to iterate and pin objects, but name could be simplified.

BrzVlad commented on mono/sgen/sgen-los.c:
```diff
@@ -63,8 +65,61 @@ struct _LOSSection {
-/* We allow read only access on the list while sweep is not running */
-LOSObject *sgen_los_object_list = NULL;
+/* We allow read only access on the array list while sweep is not running */
+static SgenArrayList sgen_los_object_array_list = SGEN_ARRAY_LIST_INIT (NULL, sgen_array_list_default_is_slot_set, sgen_array_list_default_cas_setter, INTERNAL_MEM_PIN_QUEUE);
```
We don't need to have `sgen_array_list_default_cas_setter`, put NULL instead. We don't have multiple threads racing for addition to the list since we only alloc large objects under the GC lock

lateralusX replied:
Yes, I can drop that. I looked through the code it seems that all read/write of next_index is done while holding GC lock. I believe I found one scenario that could end up changing the list without holding the GC lock, mono_gc_base_cleanup, but don't know if we could get activity into the list from other threads at that point in time. Looks like that code path could end up in major_finish_concurrent_collection, that will do a sgen_los_sweep, that could compact the los array and modify the next_slot variable.

BrzVlad replied:
We don't care about reads, we can have concurrent reads on this structure, but not writes. Only the sweeper and the mutator write to this list. If we have the mutator running while finishing a concurrent collection, then the access to this list is the last of our worries.

lateralusX replied:
We will do a stw as part of that, so that would eliminate any mutators running, but all other places we will hold the GC lock when doing sgen_perform_collection except when called from mono_gc_base_cleanup. Any reasons why we don't hold the GC lock in that scenario as well?

lateralusX replied:
So without CAS setter I see failures in sgen-new-threads-dont-join-stw.exe during shutdown (have seen this a couple of times), but only on Linux x64 Full AOT + LLVM. Maybe this PR expose an already existing issue in this area during shutdown for this specific test. Only seen on marksweep, simple and not on marksweep-conc, simple.

lateralusX commented:
Callstack for failure during shutdown of sgen-new-threads-dont-join-stw.exe on x64 Linux Full AOT + LLVM, only seen using marksweep, simple and not marksweep-conc, simple. 

```
Thread 8 (Thread 0x7f48eadd2700 (LWP 66872)):
#0  0x00007f4924236f7c in __lll_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f4924230c26 in pthread_mutex_lock () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x0000558661e2dd70 in mono_os_mutex_lock (mutex=0x5586639783a0 <loader_mutex>) at ../../mono/utils/mono-os-mutex.h:103
#3  mono_coop_mutex_lock (mutex=0x5586639783a0 <loader_mutex>) at ../../mono/utils/mono-coop-mutex.h:54
#4  mono_loader_lock () at loader.c:142
#5  0x0000558661d1999a in mono_get_seq_points (domain=domain@entry=0x558665422c30, method=0x7f48cc0045b0) at seq-points.c:270
#6  0x0000558661c8832b in mono_walk_stack_full (func=func@entry=0x558661e2d7f0 <stack_walk_adapter>, start_ctx=start_ctx@entry=0x7f48eadd13b0, domain=0x558665422c30, jit_tls=jit_tls@entry=0x7f48c4001a80, lmf=<optimized out>, lmf@entry=0x7f48eadd15e8, unwind_options=unwind_options@entry=MONO_UNWIND_LOOKUP_ALL, user_data=0x7f48eadd1580, crash_context=0) at mini-exceptions.c:1344
#7  0x0000558661c88498 in mono_walk_stack_with_ctx (func=0x558661e2d7f0 <stack_walk_adapter>, start_ctx=0x7f48eadd13b0, unwind_options=MONO_UNWIND_LOOKUP_ALL, user_data=0x7f48eadd1580) at mini-exceptions.c:1183
#8  0x0000558661e2e5d5 in mono_stack_walk (func=func@entry=0x558661e79420 <find_wrapper>, user_data=user_data@entry=0x7f48eadd15ac) at loader.c:1707
#9  0x0000558661e7c34d in is_running_protected_wrapper () at threads.c:2973
#10 mono_thread_interruption_checkpoint_request (bypass_abort_protection=0) at threads.c:5155
#11 0x00007f4920ea98d1 in wrapper_managed_to_native_object___icall_wrapper_mono_thread_interruption_checkpoint () from /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mcs/class/lib/testing_aot_full/mscorlib.dll.so
#12 0x00007f4920ea9ed8 in wrapper_managed_to_native_object___icall_wrapper_mono_threads_state_poll () from /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mcs/class/lib/testing_aot_full/mscorlib.dll.so
#13 0x00007f49206fe1d8 in mono_aot_mscorlib_icall_cold_wrapper_265 () from /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mcs/class/lib/testing_aot_full/mscorlib.dll.so
#14 0x0000000000000000 in ?? ()

Thread 7 (Thread 0x7f48eafd3700 (LWP 66871)):
#0  0x00007f4924237b5a in waitpid () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x0000558661d10f88 in dump_native_stacktrace (mctx=mctx@entry=0x7f48eafd1de0, signal=0x558662f66d77 "SIGSEGV") at mini-posix.c:1072
#2  0x0000558661d110dc in mono_dump_native_crash_info (signal=signal@entry=0x558662f66d77 "SIGSEGV", mctx=mctx@entry=0x7f48eafd1de0, info=info@entry=0x0) at mini-posix.c:1118
#3  0x0000558661c897f6 in mono_handle_native_crash (signal=signal@entry=0x558662f66d77 "SIGSEGV", mctx=mctx@entry=0x7f48eafd1de0, info=info@entry=0x0) at mini-exceptions.c:3416
#4  0x0000558661d0a3e9 in altstack_handle_and_restore (ctx=0x7f48eafd1de0, obj=0x0, flags=<optimized out>) at exceptions-amd64.c:878
#5  0x0000558661f350bc in monoeg_g_hash_table_lookup_extended (hash=<optimized out>, key=0x7f48cc004a38, orig_key=0x7f48eafd2020, value=0x7f48eafd2028) at ghashtable.c:333
#6  0x00007f48eafd2028 in ?? ()
#7  0x00000000000000a2 in ?? ()
#8  0x00000000ffff1000 in ?? ()
#9  0x00007f48eafd1ff0 in ?? ()
#10 0x00007f48cc004a38 in ?? ()
#11 0x00007f48cc004a38 in ?? ()
#12 0x0000000019800947 in ?? ()
#13 0x0000000000000000 in ?? ()

Thread 6 (Thread 0x7f48eb1d4700 (LWP 66870)):
#0  0x00007f4924236f7c in __lll_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f4924230c26 in pthread_mutex_lock () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x0000558661e2dd70 in mono_os_mutex_lock (mutex=0x5586639783a0 <loader_mutex>) at ../../mono/utils/mono-os-mutex.h:103
#3  mono_coop_mutex_lock (mutex=0x5586639783a0 <loader_mutex>) at ../../mono/utils/mono-coop-mutex.h:54
#4  mono_loader_lock () at loader.c:142
#5  0x0000558661d1999a in mono_get_seq_points (domain=domain@entry=0x558665422c30, method=0x7f48cc0045b0) at seq-points.c:270
#6  0x0000558661c8832b in mono_walk_stack_full (func=func@entry=0x558661e2d7f0 <stack_walk_adapter>, start_ctx=start_ctx@entry=0x7f48eb1d32f0, domain=0x558665422c30, jit_tls=jit_tls@entry=0x7f48cc001a80, lmf=<optimized out>, lmf@entry=0x7f48eb1d3528, unwind_options=unwind_options@entry=MONO_UNWIND_LOOKUP_ALL, user_data=0x7f48eb1d34c0, crash_context=0) at mini-exceptions.c:1344
#7  0x0000558661c88498 in mono_walk_stack_with_ctx (func=0x558661e2d7f0 <stack_walk_adapter>, start_ctx=0x7f48eb1d32f0, unwind_options=MONO_UNWIND_LOOKUP_ALL, user_data=0x7f48eb1d34c0) at mini-exceptions.c:1183
#8  0x0000558661e2e5d5 in mono_stack_walk (func=func@entry=0x558661e79420 <find_wrapper>, user_data=user_data@entry=0x7f48eb1d34ec) at loader.c:1707
#9  0x0000558661e7c34d in is_running_protected_wrapper () at threads.c:2973
#10 mono_thread_interruption_checkpoint_request (bypass_abort_protection=0) at threads.c:5155
#11 0x00007f4920ea98d1 in wrapper_managed_to_native_object___icall_wrapper_mono_thread_interruption_checkpoint () from /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mcs/class/lib/testing_aot_full/mscorlib.dll.so
#12 0x00007f4920ea9ed8 in wrapper_managed_to_native_object___icall_wrapper_mono_threads_state_poll () from /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mcs/class/lib/testing_aot_full/mscorlib.dll.so
#13 0x00007f49206fe1d8 in mono_aot_mscorlib_icall_cold_wrapper_265 () from /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mcs/class/lib/testing_aot_full/mscorlib.dll.so
#14 0x0000000000000000 in ?? ()

Thread 5 (Thread 0x7f48eb5d6700 (LWP 66868)):
#0  0x00007f4924236f7c in __lll_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f4924230c26 in pthread_mutex_lock () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x0000558661e2dd70 in mono_os_mutex_lock (mutex=0x5586639783a0 <loader_mutex>) at ../../mono/utils/mono-os-mutex.h:103
#3  mono_coop_mutex_lock (mutex=0x5586639783a0 <loader_mutex>) at ../../mono/utils/mono-coop-mutex.h:54
#4  mono_loader_lock () at loader.c:142
#5  0x0000558661d1999a in mono_get_seq_points (domain=domain@entry=0x558665422c30, method=0x7f48cc004d30) at seq-points.c:270
#6  0x0000558661c8832b in mono_walk_stack_full (func=func@entry=0x558661e2d7f0 <stack_walk_adapter>, start_ctx=start_ctx@entry=0x7f48eb5d52f0, domain=0x558665422c30, jit_tls=jit_tls@entry=0x7f48d4001a80, lmf=<optimized out>, lmf@entry=0x7f48eb5d5528, unwind_options=unwind_options@entry=MONO_UNWIND_LOOKUP_ALL, user_data=0x7f48eb5d54c0, crash_context=0) at mini-exceptions.c:1344
#7  0x0000558661c88498 in mono_walk_stack_with_ctx (func=0x558661e2d7f0 <stack_walk_adapter>, start_ctx=0x7f48eb5d52f0, unwind_options=MONO_UNWIND_LOOKUP_ALL, user_data=0x7f48eb5d54c0) at mini-exceptions.c:1183
#8  0x0000558661e2e5d5 in mono_stack_walk (func=func@entry=0x558661e79420 <find_wrapper>, user_data=user_data@entry=0x7f48eb5d54ec) at loader.c:1707
#9  0x0000558661e7c34d in is_running_protected_wrapper () at threads.c:2973
#10 mono_thread_interruption_checkpoint_request (bypass_abort_protection=0) at threads.c:5155
#11 0x00007f4920ea98d1 in wrapper_managed_to_native_object___icall_wrapper_mono_thread_interruption_checkpoint () from /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mcs/class/lib/testing_aot_full/mscorlib.dll.so
#12 0x00007f4920ea9ed8 in wrapper_managed_to_native_object___icall_wrapper_mono_threads_state_poll () from /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mcs/class/lib/testing_aot_full/mscorlib.dll.so
#13 0x00007f49206fe1d8 in mono_aot_mscorlib_icall_cold_wrapper_265 () from /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mcs/class/lib/testing_aot_full/mscorlib.dll.so
#14 0x0000000000000000 in ?? ()

Thread 4 (Thread 0x7f48eb9d8700 (LWP 66866)):
#0  0x00007f4924236f7c in __lll_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f4924230c26 in pthread_mutex_lock () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x0000558661e2dd70 in mono_os_mutex_lock (mutex=0x5586639783a0 <loader_mutex>) at ../../mono/utils/mono-os-mutex.h:103
#3  mono_coop_mutex_lock (mutex=0x5586639783a0 <loader_mutex>) at ../../mono/utils/mono-coop-mutex.h:54
#4  mono_loader_lock () at loader.c:142
#5  0x0000558661c53133 in mono_jit_compile_method_with_opt (method=method@entry=0x558665442000, opt=<optimized out>, jit_only=jit_only@entry=0, error=error@entry=0x7f48eb9d6d50) at mini-runtime.c:2477
#6  0x0000558661c5425c in mono_jit_compile_method (method=method@entry=0x558665442000, error=error@entry=0x7f48eb9d6d50) at mini-runtime.c:2495
#7  0x0000558661c8ce4c in common_call_trampoline (regs=regs@entry=0x7f48eb9d6e08, code=code@entry=0x7f49206fe1d8 <mono_aot_mscorlib_icall_cold_wrapper_265+136> "\017(\004$\017(L$\020\017(T$ \017(\\$0\017(d$@\017(l$P\017(t$`\017(|$pD\017(\204$\200", m=m@entry=0x558665442000, vt=vt@entry=0x0, vtable_slot=<optimized out>, vtable_slot@entry=0x0, error=error@entry=0x7f48eb9d6d50) at mini-trampolines.c:635
#8  0x0000558661c8d7b0 in mono_magic_trampoline (regs=0x7f48eb9d6e08, code=0x7f49206fe1d8 <mono_aot_mscorlib_icall_cold_wrapper_265+136> "\017(\004$\017(L$\020\017(T$ \017(\\$0\017(d$@\017(l$P\017(t$`\017(|$pD\017(\204$\200", arg=0x558665442000, tramp=<optimized out>) at mini-trampolines.c:773
#9  0x00007f4920ebedfa in generic_trampoline_jit () from /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mcs/class/lib/testing_aot_full/mscorlib.dll.so
#10 0x00007f49206fe1d8 in mono_aot_mscorlib_icall_cold_wrapper_265 () from /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mcs/class/lib/testing_aot_full/mscorlib.dll.so
#11 0x0000000000000000 in ?? ()

Thread 3 (Thread 0x7f492057c700 (LWP 66853)):
#0  0x00007f4924236556 in do_futex_wait.constprop () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f4924236604 in __new_sem_wait_slow.constprop.0 () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x0000558661ebd400 in mono_os_sem_wait (flags=MONO_SEM_FLAGS_ALERTABLE, sem=0x55866397bea0 <finalizer_sem>) at ../../mono/utils/mono-os-semaphore.h:203
#3  mono_coop_sem_wait (flags=MONO_SEM_FLAGS_ALERTABLE, sem=0x55866397bea0 <finalizer_sem>) at ../../mono/utils/mono-coop-semaphore.h:41
#4  finalizer_thread (unused=unused@entry=0x0) at gc.c:963
#5  0x0000558661e82893 in start_wrapper_internal (stack_ptr=<optimized out>, start_info=0x0) at threads.c:1233
#6  start_wrapper (data=0x55866546c750) at threads.c:1306
#7  0x00007f492422e4a4 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#8  0x00007f49236d3d0f in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 2 (Thread 0x7f4922fff700 (LWP 66852)):
#0  0x00007f492423417f in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x0000558661f1425b in mono_os_cond_wait (mutex=0x55866398ac00 <lock>, cond=0x55866398abc0 <work_cond>) at ../../mono/utils/mono-os-mutex.h:177
#2  get_work (job=<synthetic pointer>, do_idle=<synthetic pointer>, work_context=<synthetic pointer>, worker_index=0) at sgen-thread-pool.c:165
#3  thread_func (data=<optimized out>) at sgen-thread-pool.c:196
#4  0x00007f492422e4a4 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#5  0x00007f49236d3d0f in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 1 (Thread 0x7f4924a66780 (LWP 66837)):
#0  0x00007f492423417f in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x0000558661f172ef in mono_os_cond_wait (mutex=0x55866398adc0 <signal_mutex>, cond=0x7ffec9846e20) at mono-os-mutex.h:177
#2  mono_os_event_wait_multiple (events=events@entry=0x7ffec9846e90, nevents=nevents@entry=11, waitall=waitall@entry=1, timeout=4294967295, alertable=1) at os-event-unix.c:190
#3  0x0000558661f2bc77 in mono_thread_info_wait_multiple_handle (thread_handles=thread_handles@entry=0x7ffec9847110, nhandles=11, background_change_event=background_change_event@entry=0x0, waitall=waitall@entry=1, timeout=<optimized out>, alertable=<optimized out>) at mono-threads.c:2069
#4  0x0000558661e7a0f7 in wait_for_tids (wait=wait@entry=0x7ffec9847110, timeout=timeout@entry=4294967295, check_state_change=check_state_change@entry=0) at threads.c:3513
#5  0x0000558661e7dc4f in mono_thread_manage () at threads.c:3739
#6  0x0000558661c5fd97 in mono_main (argc=<optimized out>, argv=<optimized out>) at driver.c:2625
#7  0x0000558661c167b7 in mono_main_with_options (argv=<optimized out>, argc=<optimized out>) at main.c:52
#8  main (argc=<optimized out>, argv=<optimized out>) at main.c:434
```

BrzVlad commented:
The crash looks unfamiliar and that lane is not a good reference point. It is busted : https://jenkins.mono-project.com//job/test-mono-mainline-linux-fullaot+llvm/ and I already see that stacktrace on master there

lateralusX commented:
"Good", latest master build has the same issue and callstack, https://jenkins.mono-project.com/job/test-mono-mainline-linux-fullaot+llvm/label=debian-9-amd64/1076/parsed_console/log_content.html#WARNING1, so then this PR is not the one causing that issue.
